### PR TITLE
WP & Metalsmith set up

### PR DESCRIPTION
**Pull request by martzoukos**
The set up includes:
- A [hosted wordpress website](https://humboldtforumapi.wordpress.com/)
- Its [rest api](https://public-api.wordpress.com/rest/v1.1/sites/humboldtforumapi.wordpress.com/posts/)
- A [metalsmith](metalsmith.io) app that 1) fetches the content, 2) saves it into markdown files, 3) turns them into html files

Templating language is nunjucks.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,30 @@
...
+
+## Config
+
+`wp-config.yml` has all the configuration for mapping the wordpress json schema into local markdown files saved at the `source/data` folder
+
+
+## Development
+
+Starts server at port 8081
+```
```
Comment (HugoGiraudel) [inline README.md]: Missing empty line before.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,30 @@
...
+npm start
+```
+
+Get latest content from wordpress
+```
```
Comment (HugoGiraudel) [inline README.md]: Missing empty line before.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,30 @@
...
+npm run wp
+```
+
+Build site
+```
```
Comment (HugoGiraudel) [inline README.md]: Missing empty line before.

File: bin/build.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+const app = require('../index.js')
+
+app.build()
```
Comment (HugoGiraudel) [inline bin/build.js]: You should probably also export the app, just in case.

```
module.exports = app
```

File: bin/fetch-wp.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
+const request = require('request-promise')
+const fs = require('fs-promise')
+const path = require('path')
+const _ = require('lodash')
+
+exports.init = function(config) {
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: Missing space before opening parenthese.

Comment (HugoGiraudel) [inline bin/fetch-wp.js]: Use `module.exports`.

File: bin/fetch-wp.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+  const getFilePath = (contentItem) =>
+    path.join(config.dist_dir, `${contentItem.slug}.md`)
+
+  const wrapFrontMatter = (fm) =>
+    ['---'].concat(fm).concat('---').concat('').join('\n')
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: Could be good to store the FM delimiter in a constant.

```js
const FRONT_MATTER_DELIMITER = '---'
```

File: bin/fetch-wp.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+  const pickValues = (object) =>
+    Object.keys(config.fm_spec).reduce((acc, key) => {
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: This function should use `object`, not `config.fm_spec`. As it is right now, it’s both not pure and having a useless parameter.

File: bin/fetch-wp.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      acc[key] = _.get(object, config.fm_spec[key])
+      return acc
+    }, {})
+
+  const generateFrontMatter = (data) =>
+    Object.keys(data).reduce((acc, key) => {
+      acc.push(key + ': "' + getFrontMatterValue(data, key) + '"')
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: Any reason you force quotes here? YAML does not need quotes.

File: bin/fetch-wp.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      return acc
+    }, [])
+
+  const getFrontMatterValue = (data, key) =>
+    data[key] || config.fm_spec[key]
+
+  const createFrontMatter = (contentItem) =>
+    wrapFrontMatter(generateFrontMatter(contentItem))
+
+  const getFrontMatter = (contentItem) =>
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: I would probably compose here.

```js
const getFrontMatter = _.compose(
  wrapFrontMatter,
  generateFrontMatter,
  pickValues
);
```

File: bin/fetch-wp.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
+const request = require('request-promise')
+const fs = require('fs-promise')
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: You should probably cherry-pick what you need.

File: bin/fetch-wp.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+const path = require('path')
+const _ = require('lodash')
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: Same here.

File: bin/fetch-wp.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+  const deleteDir = (path) =>
+    fs.remove(path)
+
+  const createDir = (path) =>
+    fs.mkdir(path)
+
+  const refreshDir = (path) =>
+    deleteDir(path)
+      .then(() => createDir(path))
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: An interesting thing to do here would be to resolve again with the path so it’s accessible in any follow-up in the promise chain.

```js
const refreshDir = (path) =>
  deleteDir(path)
    .then(() => createDir(path))
    .then(() => path)
```

File: bin/fetch-wp.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+  const createPosts = (contentItems) =>
+    Promise.all(contentItems.map(createPost))
+
+  const parseResponse = (contentItems) =>
+    parseDeep(JSON.parse(contentItems))
+
+  const parseDeep = (contentItems) => {
+    if ( !Array.isArray(contentItems) ) {
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: Unnecessary spaces here.

File: bin/fetch-wp.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    deleteDir(path)
+      .then(() => createDir(path))
+
+  const createPosts = (contentItems) =>
+    Promise.all(contentItems.map(createPost))
+
+  const parseResponse = (contentItems) =>
+    parseDeep(JSON.parse(contentItems))
+
+  const parseDeep = (contentItems) => {
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: What are you trying to achieve here?

Comment (HugoGiraudel) [inline bin/fetch-wp.js]: The function seems poorly named, as no parsing seems to be done here.

File: bin/fetch-wp.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    if ( !Array.isArray(contentItems) ) {
+      return contentItems[config.response_array]
+    }
+  }
+
+  const handleError = (error) => {
+    throw new Error(error)
+  }
+
+  const refreshDist = () =>
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: I’d argue this function is not necessary.

File: bin/wp.js [hunk 16]
```diff
@@ -0,0 +1,3 @@
+const app = require('../index.js')
+
+app.wp()
```
Comment (HugoGiraudel) [inline bin/wp.js]: You should probably also export the app, just in case.

```js
module.exports = app
```

File: bin/serve.js [hunk 17]
```diff
@@ -0,0 +1,3 @@
+const app = require('../index.js')
+
+app.serve()
```
Comment (HugoGiraudel) [inline bin/serve.js]: You should probably also export the app, just in case.

```js
module.exports = app
```

File: index.js [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+const collections  = require('metalsmith-collections')
+const permalinks   = require('metalsmith-permalinks')
+const babel        = require('metalsmith-babel')
+const autoprefixer = require('metalsmith-autoprefixer')
+const sass         = require('metalsmith-sass')
+const serve        = require('metalsmith-serve')
+const watch        = require('metalsmith-watch')
+const yaml_config  = require('node-yaml-config')
+const fetchWP      = require('./bin/fetch-wp.js')
+
```
Comment (HugoGiraudel) [inline index.js]: One too many line.

File: index.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+
+const SOURCE_PATH = './source'
+const BUILD_PATH  = './_build'
+
+const config = yaml_config.load('wp-config.yml');
```
Comment (HugoGiraudel) [inline index.js]: Any reason `SOURCE_PATH` and `BUILD_PATH` are not coming from `wp-config.yml`?

File: index.js [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+const permalinksOptions = {
+  linksets: [{
```
Comment (HugoGiraudel) [inline index.js]: Any reason you use an array here?

File: index.js [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+  }]
+}
+
+const babelOptions = {
+  presets: ['es2015']
+}
+
+const watchOptions = {
+  paths: {
+    "${source}/**/*": true,
```
Comment (HugoGiraudel) [inline index.js]: Backticks if you are try to interpolate the `source` variable.

File: index.js [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+  port: 8081
+}
+
+const handleError = (err) => {
+  if (err) {
+    throw new Error(err)
+  }
+}
+
+const MetalsmithPipe = () =>
```
Comment (HugoGiraudel) [inline index.js]: This function should probably not be capitalised. Also *pipeline*  would probably make more sense.

File: index.js [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+  Metalsmith(__dirname)
+    .source(SOURCE_PATH)
+    .use(markdown())
+    .use(collections())
+    .use(permalinks(permalinksOptions))
+    .use(babel(babelOptions))
+    .use(sass())
```
Comment (HugoGiraudel) [inline index.js]: We might need `sassOptions` here.

File: index.js [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+    .use(autoprefixer())
```
Comment (HugoGiraudel) [inline index.js]: We might need `autoprefixerOptions` here.

File: index.js [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+    .use(templates('nunjucks'))
```
Comment (HugoGiraudel) [inline index.js]: We might put this in a templater option.

File: index.js [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+    .destination(BUILD_PATH)
+
+
+const buildFiles = () =>
+  MetalsmithPipe()
+    .build(handleError)
```
Comment (HugoGiraudel) [inline index.js]: Any reason you didn’t put this in the previous function directly?

File: index.js [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+const templates    = require('metalsmith-templates')
+const collections  = require('metalsmith-collections')
+const permalinks   = require('metalsmith-permalinks')
+const babel        = require('metalsmith-babel')
+const autoprefixer = require('metalsmith-autoprefixer')
+const sass         = require('metalsmith-sass')
+const serve        = require('metalsmith-serve')
+const watch        = require('metalsmith-watch')
+const yaml_config  = require('node-yaml-config')
+const fetchWP      = require('./bin/fetch-wp.js')
```
Comment (HugoGiraudel) [inline index.js]: Cherry-pick `init` here.

File: index.js [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+  MetalsmithPipe()
+    .use(serve(serveOptions))
+    .use(watch(watchOptions))
+    .build(handleError)
+
+const fetchWPContent = () => {
+  config.content_types.map((config) => fetchWP.init(config))
+}
+
+exports.build  = buildFiles;
```
Comment (HugoGiraudel) [inline index.js]: Make up your mind about semi-colons. ;)

File: package.json [hunk 29]
```diff
@@ -0,0 +1,30 @@
+{
+  "scripts": {
+    "start": "node bin/serve",
+    "wp": "node bin/wp",
```
Comment (HugoGiraudel) [inline package.json]: This one is not super explicit. Maybe `wp-import` instead?

File: package.json [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+    "build": "node bin/wp && node bin/build"
```
Comment (HugoGiraudel) [inline package.json]: You could use previous task instead.

File: source/index.md [hunk 31]
```diff
@@ -0,0 +1,5 @@
+---
+title: "Homepage"
+slug: ""
+template: home.njk
```
Comment (HugoGiraudel) [inline source/index.md]: Any reason you use a `.njk` extension rather than `.html`?

File: index.js [hunk 32]
```diff
@@ -0,0 +1,77 @@
...
+const config = yaml_config.load('wp-config.yml');
+
+const permalinksOptions = {
+  linksets: [{
+    match: { collection: 'posts' },
+    pattern: ':collection/:slug'
+  }]
+}
+
+const babelOptions = {
```
Comment (HugoGiraudel) [inline index.js]: All these options should probably be in a `build_configuration.json` or something. This file is getting too big, and JS is not a good place to store options. :)

File: bin/fetch-wp.js (hunk 5 above)
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: Also prefer exporting at the end of file.

Comment (HugoGiraudel) [inline bin/fetch-wp.js]: I would probably name this function.

File: bin/fetch-wp.js [hunk 33]
```diff
@@ -0,0 +1,79 @@
...
+const fs = require('fs-promise')
+const path = require('path')
+const _ = require('lodash')
+
+exports.init = function(config) {
+
+  const getFilePath = (contentItem) =>
+    path.join(config.dist_dir, `${contentItem.slug}.md`)
+
+  const wrapFrontMatter = (fm) =>
```
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: `fm` is not a very explicit variable name.

File: bin/fetch-wp.js (hunk 7 above)
Comment (martzoukos) [inline bin/fetch-wp.js]: `object` here should be a `contentItem` (a value of the Posts array), just like in everywhere else. Typo probably, will change.

File: bin/fetch-wp.js (hunk 8 above)
Comment (martzoukos) [inline bin/fetch-wp.js]: Carried habit from Rails projects. Can remove it.

File: bin/fetch-wp.js (hunk 14 above)
Comment (martzoukos) [inline bin/fetch-wp.js]: If the response is not an array but an object with:
```
response: {
  meta: { ... },
  data: { ... }
}
```
I'm going 1 level deep and look into the `config.response_array` property.

Comment (HugoGiraudel) [inline bin/fetch-wp.js]: Why would it look like this?

File: bin/fetch-wp.js (hunk 15 above)
Comment (martzoukos) [inline bin/fetch-wp.js]: Should I directly run the `refreshDir(config.dist_dir)` function in the program in line 73 then?
Does it not get too verbose then?

Comment (HugoGiraudel) [inline bin/fetch-wp.js]: I think it’s fine to run it directly yes.

File: bin/fetch-wp.js (hunk 11 above)
Comment (martzoukos) [inline bin/fetch-wp.js]: Do you mean something like that?
```
const { get, flowRight } = require('lodash');
```

Comment (HugoGiraudel) [inline bin/fetch-wp.js]: Yes. 

Note: please use `compose`, not `flowRight`.

File: bin/fetch-wp.js (hunk 14 above)
Comment (martzoukos) [inline bin/fetch-wp.js]: https://public-api.wordpress.com/rest/v1.1/sites/humboldtforumapi.wordpress.com/posts/

File: bin/fetch-wp.js (hunk 11 above)
Comment (martzoukos) [inline bin/fetch-wp.js]: `compose` is deprecated and doesn't work: https://lodash.com/docs

File: bin/fetch-wp.js (hunk 14 above)
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: We need to approach this a bit differently then.

File: bin/fetch-wp.js (hunk 11 above)
Comment (HugoGiraudel) [inline bin/fetch-wp.js]: You need to pull it from `lodash/fp`.

File: index.js (hunk 20 above)
Comment (martzoukos) [inline index.js]: No reason actually. Changed it 👍

File: index.js [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+    .build(handleError);
+
+const serveSite = () =>
+  MetalsmithPipe()
+    .use(serve(serveOptions))
+    .use(watch(watchOptions))
+    .build(handleError);
+
+const fetchWPContent = () => {
+  config.content_types.map((config) => fetchWP.init(config));
```
Comment (HugoGiraudel) [inline index.js]:
```js
config.content_types.map(fetchWP);
```

**Comment by martzoukos** [comment]
@HugoGiraudel ready to be reviewed! 🎉

File: .metalsmithrc [hunk 35]
```diff
@@ -0,0 +1,24 @@
...
+    pattern: posts/:slug
+  - match:
+      collection: pages
+    pattern: pages/:slug
+
+babel:
+  presets:
+  - es2015
+
+sass: {}
```
Comment (HugoGiraudel) [inline .metalsmithrc]: Don’t put empty objects in YAML.

File: .metalsmithrc [hunk 36]
```diff
@@ -0,0 +1,24 @@
...
+
+autoprefixer: {}
```
Comment (HugoGiraudel) [inline .metalsmithrc]: Don’t put empty objects in YAML.

File: README.md [hunk 37]
```diff
@@ -1,2 +1,33 @@
-# humboldt-forum
-WP Rest API with Static Site Generator on the Front End
+# CEVT
+Static Site generator, using Metalsmith. Gets content from a wordpress.com site.
```
Comment (HugoGiraudel) [inline README.md]: Missing empty line before.

File: source/index.md [hunk 38]
```diff
@@ -0,0 +1,6 @@
+---
+title: Homepage
+slug:
```
Comment (HugoGiraudel) [inline source/index.md]: Do we need to hardcode the slug?

File: bin/utils/wp-import.js [hunk 39]
```diff
@@ -0,0 +1,80 @@
...
+const { get } = require('lodash');
+
+const FRONT_MATTER_DELIMITER = '---';
+
+const importData = function (config) {
+  const getFilePath = (entityData) =>
+    path.join(getDestinationDir(), `${entityData.slug}.md`);
+
+  const wrapFrontMatter = (frontMatter) =>
+    [FRONT_MATTER_DELIMITER].concat(frontMatter).concat(FRONT_MATTER_DELIMITER).concat('').join('\n');
```
Comment (HugoGiraudel) [inline bin/utils/wp-import.js]: Worth breaking after each dot.

File: bin/utils/wp-import.js [hunk 40]
```diff
@@ -0,0 +1,80 @@
...
+const path = require('path');
+const { writeFile, remove, mkdir } = require('fs-promise');
+const { compose } = require('lodash/fp');
+const { get } = require('lodash');
+
+const FRONT_MATTER_DELIMITER = '---';
+
+const importData = function (config) {
+  const getFilePath = (entityData) =>
+    path.join(getDestinationDir(), `${entityData.slug}.md`);
```
Comment (HugoGiraudel) [inline bin/utils/wp-import.js]: Does it always have a slug?

File: bin/utils/wp-import.js [hunk 41]
```diff
@@ -0,0 +1,80 @@
...
+  const createFiles = (response) =>
+    Promise.all(response.map(createFile));
+
+  const parseResponse = (response) =>
+    getEntities(JSON.parse(response));
+
+  const getEntities = (response) =>
+    response[config.response_array] || response;
+
+  const handleError = (error) => {
```
Comment (HugoGiraudel) [inline bin/utils/wp-import.js]: Use `handleError` util instead.

File: source/scripts/application.js [hunk 42]
```diff
@@ -0,0 +1,4 @@
+const hello = (name) =>
```
Comment (HugoGiraudel) [inline source/scripts/application.js]: This can probably be removed.

File: source/styles/application.scss [hunk 43]
```diff
@@ -0,0 +1,5 @@
+$color: 'blue';
```
Comment (HugoGiraudel) [inline source/styles/application.scss]: Colors should not be quoted.

File: views/layouts/home.njk [hunk 44]
```diff
@@ -0,0 +1,16 @@
+{% include "views/partials/header.njk" %}
+
+<article>
```
Comment (HugoGiraudel) [inline views/layouts/home.njk]: This is incorrect markup.

File: .metalsmithrc (hunk 35 above)
Comment (martzoukos) [inline .metalsmithrc]: That was dumb. You're right.

File: bin/utils/wp-import.js (hunk 41 above)
Comment (martzoukos) [inline bin/utils/wp-import.js]: As we'll use https://github.com/edenspiekermann/wp-importer instead I'll leave changes on wp-importer unadressed.

**Pull request merged**